onnxruntime/core/mlas/lib/sqnbitgemm.h

```diff
@@ -0,0 +1,299 @@
+ * @param[out]  C                   Supplies the output C matrix.
+ * @param       CountM              Number of rows of A and C.
+ * @param       CountN              Number of columns of B and C.
+ * @param       CountK              Number of columns of A and rows of B.
+ * @param       lda                 Leading dimension of A.
+ * @param       BlockStrideQuantB   Number of blocks between adjacent columns of the quantized B matrix.
+ * @param       ldc                 Leading dimension of C.
+ * @param       Bias                Bias vector of length N.
+ *
+ * @return  Number of rows of A handled.
+ */
+template <size_t BlkBitWidth, size_t BlkLen, typename KernelType>
+MLAS_FORCEINLINE size_t
+MlasSQNBitGemmKernel(
+    const float* A,
+    const uint8_t* QuantBData,
+    const float* QuantBScale,
+    const uint8_t* QuantBZeroPoint,
+    float* C,
+    size_t CountM,
+    size_t CountN,
+    size_t CountK,
+    size_t lda,
+    size_t BlockStrideQuantB,
+    size_t ldc,
+    const float* Bias
+);
+
+/**
+ * @brief Dequantize B into the format expected by the Sgemm kernel.
+ *        B is block quantized and column major.
+ *        This is equivalent to dequantizing B and then running
+ *        MlasSgemmCopyPackB.
+ *
+ * @tparam BlkBitWidth  Bit width of each value in a block.
+ * @tparam BlkLen       Number of values in a block.
+ * @tparam KernelType   Hardware-specific kernel type.
+ *
+ * @param[out]  FpData              Supplies the output buffer for the dequantized B float data.
+ * @param       QuantBData          Supplies the quantized B matrix block data.
+ * @param       QuantBScale         Supplies the quantized B matrix block scale values.
+ * @param       QuantBZeroPoint     Supplies the quantized B matrix block zero point values. Optional.
+ * @param       CountN              Number of columns of B.
+ * @param       CountK              Number of rows of B.
+ * @param       BlockStrideQuantB   Number of blocks between adjacent columns of the quantized B matrix.
+ */
+template <size_t BlkBitWidth, size_t BlkLen, typename KernelType>
+MLAS_FORCEINLINE void
+MlasQNBitBlkDequantBForSgemm(
+    float* FpData,
+    const uint8_t* QuantBData,
+    const float* QuantBScale,
+    const uint8_t* QuantBZeroPoint,
+    size_t CountN,
+    size_t CountK,
+    size_t BlockStrideQuantB
+);
+
+//
+// MlasQNBitGemmOperation and helpers
+//
+
+constexpr MLAS_FORCEINLINE size_t
+MlasQNBitBlkDataSizeInBytes(size_t BlkBitWidth, size_t BlkLen)
+{
+    return BlkLen * BlkBitWidth / 8;
+}
+
+template <size_t BlkBitWidth>
+constexpr MLAS_FORCEINLINE size_t
+MlasQNBitZeroPointsForBlksSizeInBytes(size_t BlkCount)
+{
+    if constexpr (BlkBitWidth <= 4) {
+        return MlasDivRoundup(BlkCount, 2);  // 2 blocks per byte
+    } else {
+        return BlkCount;
+    }
+}
+
+MLAS_FORCEINLINE void
+MlasAddBiasForGemm(const float* Bias, float* C, size_t CountM, size_t CountN, size_t ldc)
```

Comment:
>MlasAddBiasForGemm [](http://example.com/codeflow?start=0&length=18)

it is duplicated with MlasAddBiasForGemm in q4gemm.h.

Reply:
reverted q4gemm.h change